### PR TITLE
Initial implementation of the vmm-epoll crate

### PR DESCRIPTION
**Pull request by jiangliu**
The initial implementation is derived from firecracker project, with several Traits to abstract the interfaces.
Later we will add an multi-thread safe implementation to support device hotplug.

### CODE REVIEW

**Comment by andreeaflorescu** [comment]
@jiangliu we are currently moving away from the EpollContext pattern in Firecracker. There are a few reasons for this:
1. The device implementation is split in 2 places. Looking at the block device, we have part of the emulation implemented in [devices/src/virtio/block.rs](https://github.com/firecracker-microvm/firecracker/blob/master/src/devices/src/virtio/block.rs) and the event handling is implemented in [src/vmm/lib.rs](https://github.com/firecracker-microvm/firecracker/blob/master/src/vmm/src/lib.rs). With the new EventManager we are working on, both device emulation and event handling can be implemented in the same crate. This should increase readability of the code and the general separation of concerns looks better.
2. The `MaybeHandler` and `EpollContext` are not super clear concepts and the code again it is hard to follow.

I think @acatangiu, @sandreim and @alexandruag have more details about why we don't want to use this pattern any more.

After looking a bit at this PR, it looks like it follows the same concepts that we previously had in Firecracker (and that we are trying to move away from). Should we work together on this and figure out some requirements that we have for the epoll manager before deciding what is the best path forward for all consumers of rust-vmm?

**Comment by jiangliu** [comment]
> @jiangliu we are currently moving away from the EpollContext pattern in Firecracker. There are a few reasons for this:
> 
> 1. The device implementation is split in 2 places. Looking at the block device, we have part of the emulation implemented in [devices/src/virtio/block.rs](https://github.com/firecracker-microvm/firecracker/blob/master/src/devices/src/virtio/block.rs) and the event handling is implemented in [src/vmm/lib.rs](https://github.com/firecracker-microvm/firecracker/blob/master/src/vmm/src/lib.rs). With the new EventManager we are working on, both device emulation and event handling can be implemented in the same crate. This should increase readability of the code and the general separation of concerns looks better.
> 2. The `MaybeHandler` and `EpollContext` are not super clear concepts and the code again it is hard to follow.
> 
> I think @acatangiu, @sandreim and @alexandruag have more details about why we don't want to use this pattern any more.
> 
> After looking a bit at this PR, it looks like it follows the same concepts that we previously had in Firecracker (and that we are trying to move away from). Should we work together on this and figure out some requirements that we have for the epoll manager before deciding what is the best path forward for all consumers of rust-vmm?

Absolutely needing cooperations here to unify the interface/design/implementation:)
This PR is based on the original firecracker design, with several design goals:
1) support device hotplug with new concrete implementation other than the single-thread vector version.
2) decouple epoll from the vmm/src/lib.rs and etc
3) support multiple instances of the EpollManager.
And we have disabled support of the DRIVER_UPDATE openAPI request:)

**Comment by andreeaflorescu** [comment]
@jiangliu I added a PR with the code we want to merge in Firecracker for handling events so that you can also take a look. It does not support a muti-threaded `EventManager`, but if you think that is something that you can also work it, we can iterate over that. Let me know what you think.

**Comment by jiangliu** [comment]
Replaced by #2
